### PR TITLE
[Merged by Bors] - Use a fixed state hasher in bevy_reflect for deterministic Reflect::reflect_hash() across processes

### PR DESCRIPTION
**Pull request by johanhelsing**
# Objective

- bevy_ggrs uses `reflect_hash` in order to produce checksums for its world snapshots. These checksums are sent between clients in order to detect desyncronization.
- However, since we currently use `async::AHasher` with the `std` feature, this means that hashes will always be different for different peers, even if the state is identical.
- This means bevy_ggrs needs a way to get a deterministic (fixed) hash.

## Solution

- ~~Add a feature to use `bevy_utils::FixedState` for the hasher used by bevy_reflect.~~
- Always use `bevy_utils::FixedState` for initializing the bevy_reflect hasher. 

---

## Changelog

- bevy_reflect now uses a fixed state for its hasher, which means the output of `Reflect::reflect_hash` is now deterministic across processes.

### CODE REVIEW

File: crates/bevy_reflect/Cargo.toml [hunk 1]
```diff
@@ -15,6 +15,7 @@ default = []
 bevy = ["glam", "smallvec", "bevy_math"]
 # When enabled, allows documentation comments to be accessed via reflection
 documentation = ["bevy_reflect_derive/documentation"]
+fixed_hash = []
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/Cargo.toml:18]: Comment above explaining what this does please.

**Comment by alice-i-cecile** [comment]
It's hard to imagine a case where we wouldn't want this. Why should this be a feature, rather than always enabled?

**Comment by johanhelsing** [comment]
> It's hard to imagine a case where we wouldn't want this. Why should this be a feature, rather than always enabled?

Yeah,  I guess the possibility of DOS attacks through bevy_reflect are pretty far fetched. I guess if anything it should probably an opt-in feature to make it runtime-seeded.

Hard-coding to always use fixed state

File: crates/bevy_reflect/Cargo.toml (hunk 1 above)
Comment (johanhelsing) [reply crates/bevy_reflect/Cargo.toml]: Removed it instead :)

File: crates/bevy_reflect/src/reflect.rs [hunk 2]
```diff
@@ -6,10 +6,17 @@ use crate::{
...
 
 use crate::utility::NonGenericTypeInfoCell;
-pub use bevy_utils::AHasher as ReflectHasher;
+
+use bevy_utils::FixedState;
+
+#[inline]
+pub fn reflect_hasher() -> bevy_utils::AHasher {
+    FixedState.build_hasher()
+}
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/reflect.rs:19]: Can we move this to the `utility` module?

File: crates/bevy_reflect/src/reflect.rs [hunk 3]
```diff
@@ -6,10 +6,17 @@ use crate::{
...
+    hash::BuildHasher,
 };
 
 use crate::utility::NonGenericTypeInfoCell;
-pub use bevy_utils::AHasher as ReflectHasher;
+
+use bevy_utils::FixedState;
+
+#[inline]
+pub fn reflect_hasher() -> bevy_utils::AHasher {
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/reflect.rs:17]: Nit: This should probably have a doc stating what it does, how/where it's used, and when/why to use it .

File: crates/bevy_reflect/src/reflect.rs (hunk 2 above)
Comment (johanhelsing) [reply crates/bevy_reflect/src/reflect.rs]: Done

File: crates/bevy_reflect/src/reflect.rs (hunk 3 above)
Comment (johanhelsing) [reply crates/bevy_reflect/src/reflect.rs]: Also done :)

**Comment by johanhelsing** [comment]
Sorry for the noise, still figuring out how the "reviewers" system on github works

**Review by alice-i-cecile — APPROVED** [review]

**Review by james7132 — APPROVED** [review]

**Pull request closed without merging**